### ISSUE TITLE
[RFC] - Support snapshot testing

### ISSUE THREAD

**Issue by jfairbank** [jfairbank/redux-saga-test-plan#115]
I've honestly not done a lot with snapshot testing besides React components. Would this be something worthwhile to add to Redux Saga Test Plan? Ideally, we'd serialize the yielded effects to JSON to create a snapshot.

**Comment by baldwmic** [comment]
@jfairbank what do you think of https://github.com/jimbol/expect-gen?

**Comment by jfairbank** [comment]
@baldwmic That's where I originally got the idea. I thought it might be nice to add support in Redux Saga Test Plan if people are interested in it.

**Comment by baldwmic** [comment]
I can't speak for those who use this repo as I haven't personally used it. That being said, the general approach of using snapshots that https://github.com/facebook/jest popularized has been a huge productivity boost for my team.

The redux-saga community seems fragmented on how to test sagas as demonstrated by https://github.com/redux-saga/redux-saga/issues/518. A quick search on github for "redux-saga test" returns several repos all seeming to want to accomplish a similar goal:
- https://github.com/jfairbank/redux-saga-test-plan
- https://github.com/wix/redux-saga-tester
- https://github.com/antoinejaussoin/redux-saga-testing
- https://github.com/stoeffel/redux-saga-test

It would be cool to see some cross-pollination of ideas between these maintainers, perhaps eventually resulting in one singular, best way of testing sagas. Currently, it feels like there is a great need to have good testing tools so thanks for working on this!

**Comment by jfairbank** [comment]
Thanks for the feedback! I think cross-pollination is a great idea, but I'm not sure we'll get to a singular tool. Each tool has great ideas but also opinions about API and usage, so I'm not sure if there could be a convergence. Redux Saga Test Plan is largely an API that I like to use and that may not suit others, which is totally cool. Maybe I'm wrong, though, and we'll all find a convergence that suits most everyone's needs.

**Comment by baldwmic** [comment]
Good points. As an aside, seems like the open source software model favors an evolutionary approach with maintainers pouring massive amounts of energy and time into their projects and only a few projects in a specific space gaining popularity. Perhaps this is the current best model as the phrase "competition drives innovation" implies, but I wonder if maintainers working together could generate more holistic solutions more quickly and with less waste. The main problem I see with such a collaborative approach is how to make it practical given the likely competing goals between development teams.

Regardless, it would be nice to hear from others using this project to comment on snapshot testing.

**Comment by jfairbank** [comment]
I'm finishing up work on #114, and I don't think supporting snapshot testing is going to be a big deal at all. In fact, you could _almost_ entirely snapshot the effects object that I'm going to expose in #114 already. One caveat is that you'll lose function references for effects like `call` and `fork` because Jest will serialize them to a string `[Function]`. I'm proposing a function possibly called `toJSON` that returns the same effects object but with function references converted to function names (or `<anonymous>` if they lack a name) like this `[Function someFunctionName]`.

This is an example test with the proposed API:

```js

it('serializes effects for snapshot testing', async () => {
  function* saga() {
    yield call(identity, 42);
    yield put({ type: 'FOO' });
    yield put({ type: 'BAR', payload: 'hello' });
  }

  const { toJSON } = await expectSaga(saga).run();

  expect(toJSON()).toMatchSnapshot();
});
```

And the corresponding snapshot:

```js
exports[`serializes effects for snapshot testing 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          42,
        ],
        "context": null,
        "fn": "[Function identity]",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "type": "FOO",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": "hello",
          "type": "BAR",
        },
        "channel": null,
      },
    },
  ],
}
`;
```

Anyone have any thoughts, concerns, or other ideas?

**Comment by jimbol** [comment]
Yes, this is pretty much what I'm doing with `expect-gen`. I'm glad the idea is catching on!
🌟

**Comment by jfairbank** [comment]
Not a lot of comments 😉 but this is now available in v3.1.0 via a `toJSON` method exposed in the resolved `Promise`.

```js
it('can be used with snapshot testing', () => {
  return expectSaga(saga)
    .run()
    .then((result) => {
      expect(result.toJSON()).toMatchSnapshot();
    });
});
```
